File: src/walkway_perception/scripts/lidar_reader.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
from sensor_msgs.msg import LaserScan

class LidarReader(Node):
    def __init__(self):
        super().__init__('lidar_reader')
        
        self.subscriber_ = self.create_subscription(
            LaserScan,
            '/scan',
            self.listener_callback,
            10
        )
    
    def listener_callback(self, msg):
        # LiDAR sends 360 samples, the middle index (180) is usually straight ahead.
        center_index = len(msg.ranges) // 2
        distance_ahead = msg.ranges[center_index]

        closest_obs_distance = min(msg.ranges)
        closest_obs_index = msg.ranges.index(closest_obs_distance)


        self.get_logger().info(f"Distance straight ahead: {distance_ahead:.2f} meters")
        self.get_logger().info(f"Closest obstacle distance: {closest_obs_distance:.2f} meters, at: {closest_obs_index}.")

def main(args=None):
    rclpy.init(args=args)
    node = LidarReader()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: src/walkway_perception/src/lidar_reader.cpp
#include "rclcpp/rclcpp.hpp"
#include "sensor_msgs/msg/laser_scan.hpp"
#include <algorithm> // Required for std::min_element

class LidarReader : public rclcpp::Node {
    public:
    LidarReader() : Node("lidar_reader") {
        // create_subscription<Message Type>(Topic, Queue Size, Callback FUnction)
        // std::bind is used to attach the callback function to this specific class instance
        subscription_ = this->create_subscription<sensor_msgs::msg::LaserScan>(
            "/scan", 10, std::bind(&LidarReader::listener_callback, this, std::placeholders::_1)
        );
    }

    private:
    void listener_callback(const sensor_msgs::msg::LaserScan::SharedPtr msg) const {
        int center_index = msg->ranges.size() / 2;
        float distance_ahead = msg->ranges[center_index];

        // std::min_element searches from the beginning to the end of the vector
        auto min_it = std::min_element(msg->ranges.begin(), msg->ranges.end());

        // The asterisk (*) dereferences the iterator to get the actual float value
        float closest_obs_distance = *min_it;
        int closest_obs_index = std::distance(msg->ranges.begin(), min_it);


        RCLCPP_INFO(this->get_logger(), "Distance straight ahead: %.2f meters", distance_ahead);
        RCLCPP_INFO(this->get_logger(), "Closest obstacle distance: %.2f meters, at %d", closest_obs_distance, closest_obs_index);
    }

    rclcpp::Subscription<sensor_msgs::msg::LaserScan>::SharedPtr subscription_;
};

int main(int argc, char * argv[]) {
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<LidarReader>());
    rclcpp::shutdown();
    return 0;
}